### ISSUE TITLE
Better Theme Detection in Installer and Geeklog - Better Handling of Problem Themes

### ISSUE THREAD

**Issue by eSilverStrike** [Geeklog-Core/geeklog#992]
We need to figure out a way to determine if a Geeklog theme is viable to use before Geeklog and the Installer can execute a new install, upgrade, or migration (for the installer I am talking about it checking the Geeklog theme itself, not the installer theme). If it not viable it should default back to Geeklog's default theme (if it exists). If it does not exist then Geeklog should error gracefully and the installer should instruct the Admin to either fix their theme or add the Geeklog Default theme.

Some work was done on this which generates a warning message to Admins and added version numbers: #871

We can't determine if the theme is 100% viable but we should come up with a few rules for themes moving forward. How do these sound:
- requires functions.php
- requires theme_config function in functions.php with required settings
- Themes that do not support the latest version of Geeklog usually still partially work? How far back version wise should we allow this? How do we determine this?

Installer
Installer should make sure before starting a new/upgrade/migrate:
- For new install default theme exists
- For upgrades if theme exists and if it is compatible with upgraded version of Geeklog
- Migrate does a very minimal check if current theme exists (but not if it is compatible) and switches to denim_three if it does not (but doesn't check if denim_three exists).

Installer and Geeklog should use the same code to determine default theme if existing theme is not up to the task.

**Comment by eSilverStrike** [comment]
Also another issue with theme detection is if the original Geeklog site theme is set to something that doesn't exist anymore. If allow_user_themes is set to true and the Admin user is set to another theme he may not realize that the theme config option is set to something that is not valid (so anonymous visitors for example would be getting an error but the Admin would not as he is using another theme).

We do a partial check for this in lib-common with the old professional and professional_css themes but this should be expanded to any theme set as the site theme that has problems. The Admin also should get a warning message if it is setup incorrectly (even if he is not using the theme at the time as it affects other users). We currently have a message that the Admin will see (from #871) for the current theme he is using but not if the current set site theme has problems. This message is checked and generated in the index.php file and is shown for root users on the home page.

**Comment by eSilverStrike** [comment]
Note: Issue #996 fixed a lot of the things mentioned here for Geeklog except it did not make any changes to the installer. 

- The installer should not install/upgrade/migrate if no valid theme detected.
- It would be nice if the installer could use $_CONF['theme_geeklog_default'] instead of hardcoded the value in the installer class
- It would be nice if the installer could use COM_getThemes and COM_validateTheme to determine and if needed find a valid theme

**Comment by mystralkk** [comment]
I removed all hard-coded Common::DEFAULT_THEME constant but one (at the step 1 of migration).

Since COM_getThemes and COM_validateTheme are available only after "lib-common.php" is loaded, all the installer can do is set the value of $_CONF['theme_geeklog_default'] to $_CONF['theme'] when COM_validateTheme($_CONF['theme'] returns false.  Does this make sense, @eSilverStrike ?

**Comment by eSilverStrike** [comment]
I think so... I will have to look more closely at it.
